django/db/models/query.py

```diff
@@ -45,6 +47,40 @@ def __init__(
         self.chunked_fetch = chunked_fetch
         self.chunk_size = chunk_size
 
+    async def _async_generator(self):
+        # Generators don't actually start running until the first time you call
+        # next() on them, so make the generator object in the async thread and
+        # then repeatedly dispatch to it in a sync thread.
+        sync_generator = self.__iter__()
+
+        # StopIteration cannot be raised up through sync_to_async() as too much
+        # of the Python async system was originally built on generators.
+        # Instead, capture it inside a custom sync function and change it to an
+        # exception that *can* be propagated.
+        # At some point, this could be moved to asgiref.
+        def custom_next(gen):
+            try:
+                return next(gen)
+            except StopIteration:
+                raise StopAsyncIteration()
+
+        while True:
+            try:
+                yield await sync_to_async(custom_next)(sync_generator)
```

Comment:
Aren't we worried that this will have a significant effect on performance?

[`SyncToAsync.__call__` isn't cheap at all](https://github.com/django/asgiref/blob/add576498d1e0a7eea8d70b08418d77fecd92f56/asgiref/sync.py#L348-L433) and now it will be called for every single row retrieval from a cursor which is a hot path for a lot of Django applications. Keeping this implementation around will negate any benefit async usage could have.

For backends that don't support an event based interface the only solution I can think of is a the creation of a buffer on the sync side ([e.g. 100 results which is in line with `GET_ITERATOR_CHUNK_SIZE`](https://github.com/django/django/blob/ed0a2c3238aa0b9c0d01c436d5bcd70930d696b0/django/db/models/sql/constants.py#L5-L7)) to reduce performance penalty mentioned in the comment below.

I understand that this is just preliminary work to pave the way for true async backends but the current approach here seems unusable in its current form which will encourage user to seek their own solution.

Given `QuerySet.all` fetches and caches its whole data set in memory anyway why aren't we simply making `__aiter__` the following

```python
async def __aiter__(self):
    results = await sync_to_async(list)(self.__iter__())
    yield from iter(results)
```

and either drop support for `aiterator` entirely for now or move the `_async_generator` over there with a `chunk_size` sized buffer


```python
async def aiterator(self, chunk_size=2000):
    ...
    iterator = iter(self._iterable_class(
        self, chunked_fetch=use_chunked_fetch, chunk_size=chunk_size
    ))
    while True:
        chunk = sync_to_async(lambda iterator: list(litertools.islice(iterator, chunk_size)))(iterator)
        yield from iter(chunk)
        if len(chunk) < chunk_size:
            break
```

An the buffer solution at the `BaseResult.__aiter__` layer which likely the less invasive

```python
def _list_islice(self, iterable):
    return list(litertools.islice(iterable, self.chunk_size)

async def __aiter__(self):
    iterator = iter(self)
    while True:
        chunk = sync_to_async(self._list_islice)))(iterator)
        yield from iter(chunk)
        if len(chunk) < self.chunk_size:
            break
```